phastSim/phastSim.py

```diff
@@ -607,14 +693,96 @@ def normalize(node, norm):
                 node.rate /= norm
                 if node.isTerminal:
                     node.rates /= norm
+
                 else:
                     normalize(node.belowNodes[0], norm)
                     normalize(node.belowNodes[1], norm)
 
         # normalize the rates
         normalize(node=self.genomeRoot, norm=norm)
 
+        if self.indels:
+            # if there are indels then some further calculation is require to convert the rates from 
+            # substitution rates to overall rates.
+            
+
+            def include_indels(node, scale):
+                
+                if node.isTerminal:
+                    node.insertionRate * scale
+                    node.deletionRate * scale
+                    node.rate += (node.insertionRate + node.deletionRate)
+                
+                else:
+                    node.rate = include_indels(node.belowNodes[0], scale) + include_indels(node.belowNodes[1], scale)
+
+                return node.rate
+
+            include_indels(node=self.genomeRoot, scale=scale)
+        """
+        # TODO Check this - is it right?! It's the same as the original code but seems weird. 
+        # Why are we updating self.norm to a value that is just about to get rescaled?!
+        norm = self.genomeRoot.rate / len(self.ref)
+        self.norm = norm
+
+        print("\n Total cumulative substitution rate per site before normalization: " + str(norm))
+        self.normalizeRates(self.genomeRoot, self.indels, self.codon, len(self.ref), scale)
+
+    def normalizeRates(self, rootNode, indels, codon, genome_length, scale):
```

Comment:
What is the purpose of this function, as it seems to do the same thing as normalize_rates()? 
It says below that it's for creating insertions, but I think that when you have an insertion you don't need to recalculate a normalization factor for substitution rates, you can just use the same one you calculated on line 676 (line 591 of my code) where you have "norm = norm / scale".
It would make sense to store the two values "norm" and "scale", and pass them at the function that creates rates for new inserted nodes, so that "norm" doesn't need to be recalculated.

Reply:
Ah ok - so the first original function (normalize_rates) I commented out almost all of it and all it does is call this one (normalizeRates), the only other thing it does is to set self.norm = norm. And then the rest of the code from normalize_rates is copied into normalizeRates. The reason I did this is so that the new function no longer changes any of the self variables.  
The reason why I did this is so that I thought that when creating an insertion, the rates of the insertion would need normalizing. I don't quite get why you wouldn't need to normalize the rates of the insertion, could you explain that?

Reply:
Ok, I think I get it.
So, when you create an insertion, the rates still need normalizing, but the normalizing should be the same as the one applied to the initial genome tree, that is, the normalization factors don't change as you go on with the simulations. This means that the mutation rates for the inserted nodes will be drawn from the same distributions as those of the original genome tree nodes. Does it make sense? So the only difference here would be that norm would not be calculated within normalizeRates(), but would instead be passed on as one of its inputs, and the value of norm that you calculate for the initial genome tree would then also be used for the new genome trees of inserted nodes.

Reply:
Yes, good that makes sense, I think that is what I wanted to do, hopefully I will actually have done it! :)